pyphare/pyphare/pharesee/geometry.py

```diff
@@ -0,0 +1,286 @@
+                            # in cells.
+                            if ref_pd.quantity == 'field':
+                                overlap = toFieldBox(overlap, ref_pd)
+
+                            overlaps.append({"pdatas": (ref_pd, cmp_pd),
+                                             "box": overlap,
+                                             "offset": (0, 0)})
+
+    # now dealing with first and last patches to see their patchdata overlap
+    for ref_pdname, ref_pd in patches[0].patch_datas.items():
+        for cmp_pdname, cmp_pd in patches[-1].patch_datas.items():
+
+            if cmp_pdname == ref_pdname:
+
+                gb1 = ref_pd.ghost_box
+                gb2 = cmp_pd.ghost_box
+
+                # the last patch's box needs to be
+                # shifted by -domain_box.size()
+                # indeed for a domain_box (0,64), cell 64
+                # is cell -1, so 64-65;.
+
+                offset = domain_box.size()
+                overlap = gb1 * boxm.shift(gb2, -offset)
+
+
+                if overlap is not None:
+                    if ref_pd.quantity == 'field':
+                        overlap = toFieldBox(overlap, ref_pd)
+
+                    overlaps.append({"pdatas": (ref_pd, cmp_pd),
+                                     "box": overlap,
+                                     "offset": (0, -offset)})
+
+                    # if it overlaps one way, it should overlap the other
+                    other_ovrlp = boxm.shift(gb1, offset) * gb2
+
+                    assert(other_ovrlp is not None)
+                    if ref_pd.quantity == 'field':
+                        other_ovrlp= toFieldBox(other_ovrlp, ref_pd)
+
+
+                    overlaps.append({"pdatas":(ref_pd, cmp_pd),
+                                     "box" : other_ovrlp,
+                                     "offset":(offset, 0)})
+
+
+    return overlaps
+
+
+
+
+def hierarchy_overlaps(hierarchy):
+    """
+    returns all overlaps for the given hierarchy
+    """
+    overlaps = {}
+    for ilvl, lvl in enumerate(hierarchy.patch_levels):
+        overlaps[ilvl] = compute_overlaps(lvl.patches, hierarchy.refined_domain_box(ilvl))
+    return overlaps
+
+
+
+def get_periodic_list(patches, domain_box):
+    """
+    given a list of patches and a domain box the function
+    returns a list of patches sorted by origin where the
+    first (resp. last) patch is the last (resp. first) patch
+    of the sorted list, if that patch touches the upper (resp. lower)
+    border of the domain
+    """
+    from copy import copy
+
+    sorted_patches = sorted(patches, key=lambda p: p.origin)
+
+    if touch_domain_border(sorted_patches[-1].box, domain_box, "upper"):
+        last_patch = copy(sorted_patches[-1])
+        shift_patch(last_patch, -domain_box.upper)
+        sorted_patches.insert(0, last_patch)
+
+    if touch_domain_border(sorted_patches[0].box, domain_box, "lower"):
```

Comment:
Not sure I understand the question.
Do you mean, say, if only the -1 patch touches border and not the 0 ?
we'd still need to shift -1 to the left side and add it to the patch list, no ? why not ?

Reply:
> we'd still need to shift -1 

indeed why not, if only on patch exists is there even periodicity.

Reply:
on a refined level that has plenty of patches, you could have one that touches the upper border and no patch in that level that touches the lower border, and vice versa. You'd then add only the one that touches to the list.
what's the problem ?

Reply:
no problem, just a question of "what's the point"
If there periodic levelghost overlap then it makes sense I suppose, otherwise I don't see what use it is.
are field overlaps valid if it's just ghostbox overlaps without domain?

Reply:
ok so the purpose of this function is just to see if a patch touches a periodic border and if so, put a copy on the other side. In what context and how this is used is not relevant in the scope of this function I think.

Now when used for overlap calculation: it is useful to have a copy of the border patch shifted on the other side to not forget the part of the patch that overlaps the nodes (ghosts or domain) of the other side.

In this PR overlaps are just the intersection of two PatchData ghost boxes. For instance for field : 


```
                                        vvv
   ooooo-------------ooooo
                                        ooooo----------------ooooo

```


where - and o represent domain and ghost cells, respectively, 

this is an overlap even though no ghost node overlaps a domain node.

In the simulation, the ghost nodes are filled from neighbour domain nodes, but in the test we can ignore that and just check that nodes of cells having equal AMR index should have equal values, and cells of equal AMR index should have the same particle content, no matter how this content was obtained.